core/commands/files/files.go

```diff
@@ -0,0 +1,555 @@
+			return
+		}
+
+		res.SetOutput(&Object{
+			Hash: k.B58String(),
+		})
+	},
+	Marshalers: cmds.MarshalerMap{
+		cmds.Text: func(res cmds.Response) (io.Reader, error) {
+			out := res.Output().(*Object)
+			return strings.NewReader(out.Hash), nil
+		},
+	},
+	Type: Object{},
+}
+
+var FilesCpCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "copy files into mfs",
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("src", true, false, "source object to copy"),
+		cmds.StringArg("dest", true, false, "destination to copy object to"),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		node, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		src := req.Arguments()[0]
+		dst := req.Arguments()[1]
+
+		var nd *dag.Node
+		switch {
+		case strings.HasPrefix(src, "/ipfs/"):
+			p, err := path.ParsePath(src)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			obj, err := core.Resolve(req.Context(), node, p)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			nd = obj
+		default:
+			fsn, err := mfs.Lookup(node.FilesRoot, src)
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			obj, err := fsn.GetNode()
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+
+			nd = obj
+		}
+
+		err = mfs.PutNode(node.FilesRoot, dst, nd)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+	},
+}
+
+type Object struct {
+	Hash string
+}
+
+type FilesLsOutput struct {
+	Entries []mfs.NodeListing
+}
+
+var FilesLsCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "List directories",
+		ShortDescription: `
+List directories.
+
+Examples:
+
+    $ ipfs files ls /welcome/docs/
+    about
+    contact
+    help
+    quick-start
+    readme
+    security-notes
+
+    $ ipfs files ls /myfiles/a/b/c/d
+    foo
+    bar
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.StringArg("path", true, false, "path to show listing for"),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption("l", "use long listing format"),
+	},
+	Run: func(req cmds.Request, res cmds.Response) {
+		path := req.Arguments()[0]
+		nd, err := req.InvocContext().GetNode()
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		fsn, err := mfs.Lookup(nd.FilesRoot, path)
+		if err != nil {
+			res.SetError(err, cmds.ErrNormal)
+			return
+		}
+
+		switch fsn := fsn.(type) {
+		case *mfs.Directory:
+			listing, err := fsn.List()
+			if err != nil {
+				res.SetError(err, cmds.ErrNormal)
+				return
+			}
+			res.SetOutput(&FilesLsOutput{listing})
+			return
+		case *mfs.File:
+			parts := strings.Split(path, "/")
+			name := parts[len(parts)-1]
+			out := &FilesLsOutput{[]mfs.NodeListing{mfs.NodeListing{Name: name, Type: 1}}}
+			res.SetOutput(out)
+			return
+		default:
+			res.SetError(errors.New("unrecognized type"), cmds.ErrNormal)
+		}
```

Comment:
i'd put this (the meat) `ls` into the `mfs` module (or a subpkg like `mfs/cmds` or something. i like how @rht has been moving add and cat into core. these `Commands` should really just be glue between the `commands` library, and the functionality. (i.e. so that users who import IPFS as a library could use them)